Specify additional configuration directories using command arguments

paulhopkins opened the pull request:
This adds a global command line option to specify additional configuration directories. The argument either takes a directory path, or a combination of package and sub-directory. This form fits in with suggestions from @citibeth for package level `package.yaml` files. Examples:

```
$ spack --config path/to/config/dir install myapp
==> Adding configuration scope 'config/dir'
...
$ spack --config myapplication:dir install myapp
==> Adding configuration scope 'application:dir'
...
```

Some juggling of when to add sub-command arguments was required so that they were aware of the new scopes. Note, that the short form of this argument, `-c`, conflicts with the `--color` command added by me in #3013. On reflection this command should claim the `-c` form, and the other command only use `--color`, `--no-color`, though I am open to suggestions.

There are no tests associated with this PR, but I could write a test suite that repeatedly calls spack using `subprocess` and checks the output, if that is acceptable.

citibeth commented:
Have you seen #2686?  Sorry for the duplication of effort.

citibeth commented:
I would be interested if there are any differences in the approaches we took implementing this feature --- and which one might be better in the end.

Also please read up on #2694 and #2700.

We are currently holding off on merging this feature, pending a more encompassing design for Spack Environments.  Would you be interested in participating in the Spack Environments Working Group?

luigi-calori commented:
I would also be interested in this approach as we would like to version all the deployment phases on different clusters. This PR look similar to #2686 ... I would like to test one ... any hints on which is more appropriate to test?
@tgamblin

tgamblin commented:
This is mostly the same as #2686, but it adds the ability to have multiple named package-level configuration directories.  It seems like there's some consensus and need for the top-level `--config` option.  I am not sure about the per-package configs and would like to hold off on that as we figure out how to integrate environment support in Spack.

@citibeth: what do you think of the per-package configs?  I'd rather think through the use cases for those before proceeding on that.

But it seems reasonable to integrate the top-level `--config` now -- I don't think that interferes or overcomplicates things and might actually make it easier to debug `packages.yaml` issues.  @luigi-calori: would that be enough for your use case?

Thoughts?

citibeth commented:
> On Tue, Feb 14, 2017 at 12:37 PM, Todd Gamblin <notifications@github.com> wrote:
> This is mostly the same as #2686, but it adds the ability to have multiple named package-level configuration directories. It seems like there's some consensus and need for the top-level --config option. I am not sure about the per-package configs and would like to hold off on that as we figure out how to integrate environment support in Spack.

I agree with the assessment.

A small/technical difference is that this PR loads additional scopes in `bin/spack` and #2686 loads them in `config.py` where other scopes are loaded.  At the risk of baselessly supporting my own design, I will suggest that the placement of the code in #2686 seems more logical/foolproof to me.  But I also see that placing the code there doesn't support per-package configs.
 
> @citibeth: what do you think of the per-package configs? I'd rather think through the use cases for those before proceeding on that.

A config that's only loaded for a single package sounds similar in functionality to a `BundlePackage`; see #3133, and especially discussion in #3134.

Although command-line scopes have served me well, I've also seen a lot of resistance to putting stuff in `packages.yaml`; and I'm not convinced that adding command-line scopes will fix that problem.  People want to put everything they need into packages, not these extra YAML files.  I believe reasons are:

1. `packages.yaml` is a separate file with different syntax and semantics from `package.py`.  People don't want to learn yet another abstraction.

2. People want to ship a configuration of their system (usually including multiple packages) that "just works" as part of Spack, out-of-the-box.  As of now, we have no mechanism for including `packages.yaml` files in the Spack distro.

For these reasons, I want to see how much functionality I currently get from command-line scopes can be shifted to `BundlePackage`s instead.  With that transition, I'm wondering if command-line scopes will turn out to be less useful than we currently believe; I don't know for sure, but I'm wondering.  I *do* think we should hold off on per-package command-line scopes, since they seem to work even more like `BundlePackage`, which I think is a paradigm that users will ultimately like better.

> But it seems reasonable to integrate the top-level --config now -- I don't think that interferes or overcomplicates things and might actually make it easier to debug packages.yaml issues.

Be warned that as soon as you do that, people will run into underlying bugs involved in merging scopes; see #2694.  Two bugs in particular:

1. A small bug that is fixed in #2694 
2. Larger problems involving merging lists of variants properly.  I believe this requires more thought, which I have not put into it.  For the simple cases I use it for, it does work --- as long as you consitently use the correct list (bracketed) syntax in specifying variants.

For this reason, I would consider command-line scopes to be a Beta feature at this point --- useful, but not without pitfalls.

> @luigi-calori: would that be enough for your use case?

I would be interested in seeing the extent to which `BundlePackage` can solve current problems for @luigi-calori and @paulhopkins

luigi-calori commented:
I was thinking to use just one single config folder for each cluster. I' m trying to version the full deployment recipes, so would like to keep config folder versioned.
I currently manage that with a custom script that is cumbersome.
Being able to specify a config folder on the command line would help in keeping that info versioned.

citibeth commented:
It sounds to me like you will probably want to use some combination of
`BundlePackage` and per-cluster config files.5

On Tue, Feb 14, 2017 at 1:31 PM, Luigi Calori <notifications@github.com>
wrote:

> I was thinking to use just one single config folder for each cluster. I' m
> trying to version the full deployment recipes, so would like to keep config
> folder versioned.
> I currently manage that with a custom script that is cumbersome.
> Being able to specify a config folder on the command line would help in
> keeping that info versioned.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3142#issuecomment-279793552>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd8bgsklxDRje_pslu3VzLHdrqF6Pks5rcfL_gaJpZM4MAZ8m>
> .
>

luigi-calori commented:
Yes, likely some cluster independent bundle packages and one config dir for each cluster

paulhopkins commented:
I am sorry, I did not see #2686. In my humble opinion I think my approach is a bit cleaner as all argument parsing is kept within the bin/spack and cmd/__init__.py files rather than in the core Spack library. I think this is especially important if Spack code were accessed as a library. The only downside of this is that I can't be sure that some config has not already been read and fixed by the time the command line is parsed and new scopes are added. This could be alleviated by lazy loading of config, for example `pkgsort` is now loaded only when required. 

My approach does have the ability to specify package level configuration files. My reason for including this is to have a mechanism for recommending a particular package version, but not explicitly requiring it. I am already using hierarchical metapackages to manage the installation of many applications, which also specify their dependencies hierarchically . Without a `packages.yaml` file each top level package would have to `depend_on` every package in the dependency tree. The `packages.yaml` can safely contain more packages than are required, and can therefore be shared, e.g. using a top level suggestions file to install a small part of the tree only.

The downside of this is that one could have used named variants to describe the dependency tree that was used, but I don't know if that is the Spack way, since variants or other parameters could be set by the user or other packages, and the dependency trees are normally described by the hash only.

If the syntax is a problem then would it be better to have a Python file with `suggests('package@version+variant') directives?

Two design decisions I wasn't sure about:
  * Translating long paths into a scope name; I assumed that the package level paths would be short, but if somebody specifies a normal path that is very long I use the last two elements. I see that you use the last part only. Neither approach checks to make sure they are unique.
  * Whether to automatically include a platform subdirectory with each custom directory. In your approach the documents say that you do, but I am not sure the code agrees.

Neither PRs add any tests, which I think need to be performed by actually running from a command line and checking the output, rather than running internal functions.

Finally, yes I would like to participate in the Spack Environments Working Group.

citibeth commented:
> My approach does have the ability to specify package level configuration files. My reason for including this is to have a mechanism for recommending a particular package version, but not explicitly requiring it...

I see, I had considered it possible that someone would want to do that.  I think the best way would be to add a weak form of `depends_on()`.  For example: MyPackage `depends_on('netcdf', version='4.1', weak=True)`.  The concretizer would include `netcdf@4.1` in the DAG, but *only* if a non-week `depends_on()` is found.

Unfortunately, we don't have `depends_on(weak=True)` (although it would be great if such a PR could be submitted without too much effort, @tgamblin???).  In its absence, thought, I think that the flaws you pointed out are not fatal.  The only problem if a top-level bundle packaged depends on something unnecessary, is that unnecessary package will be built.  It will not actually be linked into any binaries, and everything will still work.

> The downside of this is that one could have used named variants to describe the dependency tree that was used, but I don't know if that is the Spack way, since variants or other parameters could be set by the user or other packages, and the dependency trees are normally described by the hash only.

??? Not sure what you're saying.

>   * Translating long paths into a scope name; I assumed that the package level paths would be short, but if somebody specifies a normal path that is very long I use the last two elements. I see that you use the last part only. Neither approach checks to make sure they are unique.

Is the scope name important?  What happens if a scope is unnamed?

>   * Whether to automatically include a platform subdirectory with each custom directory. In your approach the documents say that you do, but I am not sure the code agrees.

No no, please do not.  I suppose my docs are wrong then?

> Neither PRs add any tests, which I think need to be performed by actually running from a command line and checking the output, rather than running internal functions.

We would test it by assembling the final scope and comparing to what we know it should be.  That is known to be buggy right now.

citibeth commented:
> yes I would like to participate in the Spack Environments Working Group.

Please email @tgamblin and he will get you on the list.  Also see here:

https://github.com/LLNL/spack/wiki/Spack-Environments-WG

citibeth commented:
@paulhopkins Would you be interested in seeing this PR through to conclusion?  Although I don't expect to personally use per-package scopes, there is no reason not to include the feature that at least one person wants.  Maybe Spack will get to the point in the future that we feel this feature is redundant and can remove it; but that should not prevent us from merging and using per-package scopes now.

OTOH, the documentation on #2686 is clearly more extensive (although I haven't read it recently).  Would you be able to take the best of the documentation from #2686, merge it into this PR, and then update as needed?  I think we'd then be really close to getting this merged.

paulhopkins commented:
@citibeth I will pick this up next week by merging the documentation from #2686, and addressing the comments here. I think that something like `depends_on('netcdf', version='4.1', weak=True)` or maybe just `prefers('netcdf@4.1')` (is this better than suggests?) would be useful, but I still want something that could be used on sub-packages. Many thanks for the input.

citibeth commented:
> I think that something like depends_on('netcdf', version='4.1', weak=True)
> or maybe just prefers('netcdf@4.1') (is this better than suggests?)
>

I'm not married to either syntax; we should probably solicit further
opinions.  Either way, I don't know how to implement it, and believe that
implementation would require more substantial change than this PR or
#2686.  Therefore, I think that weak references should be left for another
PR.


> would be useful, but I still want something that could be used on
> sub-packages.
>

Not sure what you mean by that?

alalazo commented:
This can be done since a while with:
```console
spack -C <my-config-dir> <command>
```
